docs/find-duplicate-skus/script.liquid

```diff
@@ -0,0 +1,103 @@
+{% assign mappings = hash %}
+{% assign dups = hash %}
+{% assign cursor = nil %}
+
+{% for i in (0..500) %}
+  {% assign after = i | times: 250 %}
+  {% capture query %}
+    query {
+      productVariants(
+        first: 250
+        after: {{ cursor | json }}
+      ) {
+        pageInfo {
+          hasNextPage
+        }
+        edges {
+          node {
+            id
+            sku
+            inventoryItem {
+              tracked
+            }
+          }
+          cursor
+        }
+      }
+    }
+  {% endcapture %}
+
+  {% assign result = query | shopify %}
+
+  {% if event.preview %}
+    {% capture result_json %}
+      {
+        "data": {
+          "productVariants": {
+            "edges": [
+              {
+                "node": {
+                  "id": "gid://shopify/ProductVariant/1234567890",
+                  "sku": "sku123",
+                  "inventoryItem": {
+                    "tracked": "true"
+                  }
+                }
+              },
+              {
+                "node": {
+                  "id": "gid://shopify/ProductVariant/123456789",
+                  "sku": "sku123",
+                  "inventoryItem": {
+                    "tracked": "true"
+                  }
+                }
+              }
+            ]
+          }
+        }
+      }
+    {% endcapture %}
+
+    {% assign result = result_json | parse_json %}
+  {% endif %}
+  
+  {% assign cursor = result.data.productVariants.edges.last.cursor %}
+  {% assign productVariants = result.data.productVariants.edges | map: "node" %}
+
+  {% for pv in productVariants %}
+    {% if pv.sku == null or pv.sku == empty %}
+      {% continue %}
+    {% endif %}
+
+    {% if options.exclude_inventory_not_tracked__boolean == true %}
+      {% if pv.inventoryItem.tracked == false %}
+        {% continue %}
+      {% endif %}
+    {% endif %}
+
+    {% assign existingPv = mappings[pv.sku] %}
+
+    {% if existingPv %}
+      {% if existingPv != pv.id %}
+        {% assign dups[pv.sku] = true %}
+      {% endif %}
+    {% else %}
+      {% assign mappings[pv.sku] = pv.id %}
+    {% endif %}
+  {% endfor %}
+
+  {% unless result.data.productVariants.pageInfo.hasNextPage %}
+    {% break %}
+  {% endunless %}
+{% endfor %}
+
+{% assign skus = productVariants | map: "sku" %}
+{% assign uniqskus = skus | uniq %}
+
+{% if skus.size == uniqskus.size %}
+  {% comment %}No exact duplicate SKUS {% endcomment %}
+  {% break %}
+{% endif %}
+
+{% action "echo" dups %}
```

Comment:
This ends up logging a hash - the keys are valuable (since they're the SKUs), but the values are all just `true`.

If we're using a hash, I want the values to be useful - maybe a _count_ of how many times the SKU was encountered? Or a list of product and variant IDs, maybe?

_Or_, if that's too much, let's just return an array of SKUs instead. This could be done by initializing `dups` to an _array_ instead (`{% assign dups = array %}`, at the top), adding to it as duplicates are found, making sure not to add it if there's already a copy of that SKU in the array:

```liquid
{% unless dups contains pv.sku %}
  {% assign dups = dups | push: pv.sku %}
{% endunless %}
```

☝️ This example uses the `push` filter, which is _so_ new it's not even documented, but you can read more about it in this thread:

https://usemechanic.slack.com/archives/C01KF3B4PUK/p1618346646161700?thread_ts=1616710753.074600&cid=C01KF3B4PUK